libs/SmartSync/SmartSync/Classes/Util/SFRefreshSyncDownTarget.m

```diff
@@ -0,0 +1,259 @@
+    SFRefreshSyncDownTarget* syncTarget = [[SFRefreshSyncDownTarget alloc] init];
+    syncTarget.queryType = SFSyncDownTargetQueryTypeRefresh;
+    syncTarget.soupName = soupName;
+    syncTarget.objectType = objectType;
+    syncTarget.fieldlist = fieldlist;
+    syncTarget.countIdsPerSoql = 500;
+    return syncTarget;
+}
+
+#pragma mark - To dictionary
+
+- (NSMutableDictionary*) asDict {
+    NSMutableDictionary *dict = [super asDict];
+    dict[kSFSyncTargetRefreshSoupName] = self.soupName;
+    dict[kSFSyncTargetRefreshObjectType] = self.objectType;
+    dict[kSFSyncTargetRefreshFieldlist] = self.fieldlist;
+    dict[kSFSyncTargetRefreshCountIdsPerSoql] = [NSNumber numberWithUnsignedInteger:self.countIdsPerSoql];
+    return dict;
+}
+
+# pragma mark - Data fetching
+
+- (void) startFetch:(SFSmartSyncSyncManager*)syncManager
+       maxTimeStamp:(long long)maxTimeStamp
+         errorBlock:(SFSyncDownTargetFetchErrorBlock)errorBlock
+      completeBlock:(SFSyncDownTargetFetchCompleteBlock)completeBlock {
+
+    // During reSync, we can't make use of the maxTimeStamp that was captured during last refresh
+    // since we expect records to have been fetched from the server and written to the soup directly outside a sync down operation
+    // Instead during a reSymc, we compute maxTimeStamp from the records in the soup
+    self.isResync = maxTimeStamp > 0;
+    [self getIdsFromSmartStoreAndFetchFromServer:syncManager
+                                      errorBlock:errorBlock
+                                   completeBlock:completeBlock];
+}
+
+- (void) continueFetch:(SFSmartSyncSyncManager*)syncManager
+            errorBlock:(SFSyncDownTargetFetchErrorBlock)errorBlock
+         completeBlock:(SFSyncDownTargetFetchCompleteBlock)completeBlock {
+    if (self.page > 0) {
+        [self getIdsFromSmartStoreAndFetchFromServer:syncManager
+                                          errorBlock:errorBlock
+                                       completeBlock:completeBlock];
+    }
+    else {
+        completeBlock(nil);
+    }
+}
+
+- (void) getListOfRemoteIds:(SFSmartSyncSyncManager*)syncManager
+                   localIds:(NSArray*)localIds
+                 errorBlock:(SFSyncDownTargetFetchErrorBlock)errorBlock
+              completeBlock:(SFSyncDownTargetFetchCompleteBlock)completeBlock {
+    if (localIds == nil) {
+        completeBlock(nil);
+        return;
+    }
+    
+    __block NSMutableArray* remoteIds = [NSMutableArray new];
+    __block NSUInteger sliceSize = self.countIdsPerSoql;
+    __block NSUInteger countSlices = ceil((float)localIds.count / sliceSize);
+    __block NSUInteger slice = 0;
+    __block NSString* idFieldName = self.idFieldName;
+    __block SFSyncDownTargetFetchCompleteBlock fetchBlockRecurse = ^(NSArray *records) {};
+    SFSyncDownTargetFetchCompleteBlock fetchBlock = ^(NSArray* records) {
+        // NB with the recursive block, using weakSelf doesn't work (it goes to nil)
```

Comment:
This can/will happen with `__weak` if the owning container object—in this case, the instance of `SFRefreshSyncDownTarget`—gets released before one or more self references inside the block are dereferenced.  This is expected behavior, as classes aren't guaranteed to stay around inside a block, if the proper reference doesn't exist.

Weakifying/strongifying `self` inside of blocks is generally only necessary if the block *itself* will be retained on its owning object container.  That's not the case here.  But consider:

```obj-c
@interface BlockOwningClass ()

@property (nonatomic, copy) void (^myCopiedBlock)();
@property (nonatomic, copy) NSString *myString;

@end

@implementation BlockOwningClass

- (void)setAndCallMyBlock {
    self.myString = @"A string";
    self.myCopiedBlock = ^{
        NSLog(@"self.myString: %@", self.myString);
    }
    self.myCopiedBlock();
}

@end
```

This is likely to force a retain cycle, because the ownership of the block itself bumps the retain count, then the block captures `self`.

So the short answer here is, "capturing" `self` is okay in this case, and if any issue exists, it's possibly the expectation that the parent class should still be around/retained (somewhere) at the time this block is being called?